src/core/public/application/application_service.tsx

```diff
@@ -17,159 +17,154 @@
       },
-      registerLegacyApp: (app: LegacyApp) => {
-        if (this.legacyApps$.value.has(app.id)) {
-          throw new Error(`A legacy application is already registered with the id "${app.id}"`);
-        }
-        if (this.legacyApps$.isStopped) {
+      registerLegacyApp: app => {
+        const appBasePath = `/app/${app.id.split(':')[0]}`;
+
+        if (this.registrationClosed) {
           throw new Error(`Applications cannot be registered after "setup"`);
+        } else if (this.legacyApps.has(app.id)) {
+          throw new Error(`A legacy application is already registered with the id "${app.id}"`);
+        } else if (hasAppBasePath(this.mounters, appBasePath)) {
+          throw new Error(
+            `An application is already registered with the appBasePath "${appBasePath}"`
+          );
         }
 
-        this.legacyApps$.next(new Map([...this.legacyApps$.value.entries(), [app.id, app]]));
+        const mount: LegacyAppMounter = () => redirectTo(basePath.prepend(appBasePath));
+        this.legacyApps.set(app.id, app);
+        this.mounters.set(app.id, { appBasePath, mount, unmountBeforeMounting: true });
       },
-      registerMountContext: this.mountContext!.registerContext,
     };
   }
 
-  public async start({
-    http,
-    injectedMetadata,
-    redirectTo = (path: string) => (window.location.href = path),
-  }: StartDeps): Promise<InternalApplicationStart> {
+  public async start({ injectedMetadata }: StartDeps): Promise<InternalApplicationStart> {
     if (!this.mountContext) {
-      throw new Error(`ApplicationService#setup() must be invoked before start.`);
+      throw new Error('ApplicationService#setup() must be invoked before start.');
     }
 
-    // Disable registration of new applications
-    this.apps$.complete();
-    this.legacyApps$.complete();
-
-    const legacyMode = injectedMetadata.getLegacyMode();
-    const currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
-    const { availableApps, availableLegacyApps, capabilities } = await this.capabilities.start({
-      apps: new Map([...this.apps$.value].map(([id, { app }]) => [id, app])),
-      legacyApps: this.legacyApps$.value,
-      injectedMetadata,
-    });
-
-    // Only setup history if we're not in legacy mode
-    const history = legacyMode ? null : createBrowserHistory({ basename: http.basePath.get() });
+    this.registrationClosed = true;
+    this.currentAppId$ = new BehaviorSubject<string | undefined>(undefined);
+    const { capabilities } = await this.capabilities.start({ injectedMetadata });
+    const availableMounters = filterAvailable(this.mounters, capabilities);
 
     return {
-      availableApps,
-      availableLegacyApps,
+      availableApps: filterAvailable(this.apps, capabilities),
+      availableLegacyApps: filterAvailable(this.legacyApps, capabilities),
       capabilities,
+      currentAppId$: this.currentAppId$,
```

Comment:
Should expose Observable instead. Otherwise, `.next` can be called from the outside.

Reply:
This is an internal API, and was already a `Subject`. @joshdover thoughts on switching this to only expose the observable now?

Reply:
Yeah makes sense to only expose the Observable